### DISCUSSION TITLE
How-to for FUSE device mount inside container with Singularity 4 and OCI mode

### DISCUSSION THREAD

**Discussion by marcodelapierre** [sylabs/singularity#2319] [Q&A] [answered]
Hi @dtrudg ,

we actually have a current topic where we might benefit from some advice from you folks: we need to be able to mount `/dev/fuse`  inside the container.

Quoting my colleague @jordeu from an internal chat:

I understood that when you enable OCI mode on singularity 4.0 then they also use the standard [CDI](https://docs.sylabs.io/guides/latest/user-guide/oci_runtime.html?_gl=1*1qj989h*_gcl_au*MjExNTM3Njk2MS4xNjk2NDMxODQ5*_ga*NDIxMjcyMzM0LjE2OTY0MzE4NDk.*_ga_X710KLJKK6*MTY5OTI1MzI4Ni4yLjEuMTY5OTI1MzI4Ni4wLjAuMA..#container-device-interface-cdi) to mount devices.

This is something similar to our [Kubernetes Device Plugin](https://github.com/nextflow-io/k8s-fuse-plugin), but not exactly the same.

So, we need to figure out how to create/install this CDI driver and then it should be possible to run with something like:
```
singularity exec  --device nextflow.io/fuse=1 --oci [..] <img> <cmd>
```

Do you have any suggestions/guidelines/tips in this space?
What would be the smoothest way to bind mount a FUSE device in a OCI container with Singularity V4?

Thank you,

**Comment by marcodelapierre** [comment]
Actually, found this straight after writing this issue:

would the [`--fusemount`](https://docs.sylabs.io/guides/latest/user-guide/bind_paths_and_mounts.html#fuse-mounts) option be the best fit ?

**Comment by dtrudg** [comment]
Hi @marcodelapierre - converted to a discussion as this is a question rather than an issue.

If you use Singularity in non-oci mode then the default is to bind all of `/dev` so you get `/dev/fuse` there...

```
singularity run ubuntu_latest.sif 
Singularity> ls /dev/fuse
/dev/fuse
```

However, you probably won't be able to do a lot with it, unless you also add `--fakeroot` to have the container running in a user namespace with a root-user mapping and capabilities granted inside the user namespace.

In OCI-mode then you could use CDI, and you could also do a bind...

```
 $ singularity run -B /dev/fuse --oci docker://ubuntu 
INFO:    Using cached OCI-SIF image
dtrudg-sylabs@mini:~$ ls -lah /dev/fuse
crw-rw-rw-. 1 nobody nogroup 10, 229 Nov  6 11:14 /dev/fuse
```

The `--fusemount` flag is a something a bit different. It is for the native mode (non-OCI-mode). It creates a specific FUSE mount associated with a file descriptor at the source end... allowing the FUSE binary providing the fs to then be run on the host or from the container.

What's the overall goal here?

**Comment by marcodelapierre** [comment]
Thanks for the quick reply Dave.

@jordeu can you have a look here, maybe we can identify a good solution for Fusion vs Singularity OCI in this context?

**Comment by jordeu** [comment]
Hi @dtrudg, we are just trying to mount a FUSE tool inside the container as we were doing before but now with OCI mode enabled.

Here a reproducible example of what we do without OCI:

1. Download gocryptfs on the host
```
$ wget https://github.com/rfjakob/gocryptfs/releases/download/v2.4.0/gocryptfs_v2.4.0_linux-static_amd64.tar.gz
$ tar -xvzf gocryptfs_v2.4.0_linux-static_amd64.tar.gz
```

2. Run singularity with fake root but without OCI, and mount a gocryptfs volume.
```
$ singularity run --fakeroot docker://ubuntu

Singularity> mkdir cipher_nonoci plain_nonoci
Singularity> ./gocryptfs -init cipher_nonoci
Singularity> ./gocryptfs cipher_nonoci plain_nonoci
Singularity> mount | grep cipher
/home/jordi/fusetest/cipher_nonoci on /home/jordi/fusetest/plain_nonoci type fuse.gocryptfs (rw,nosuid,nodev,relatime,user_id=0,group_id=0,max_read=1048576)
```

3. Now try the same but with OCI enabled
```
$ singularity run --fakeroot -B $(pwd):/root -B /dev/fuse --oci docker://ubuntu

Singularity> mkdir cipher_oci plain_oci
Singularity> ./gocryptfs -init cipher_oci
Singularity> ./gocryptfs cipher_oci plain_oci

fs.Mount failed: exec: "/bin/fusermount": stat /bin/fusermount: no such file or directory
```

It is always failing to mount. Is there any seccomp profile that blocks mounting enabled when running on OCI mode?

**Reply by dtrudg** [reply]
seccomp shouldn't be involved... but the capability set in OCI-mode is different, to mirror Docker. I would guess that in this case, `gocryptfs` needs a capability (likely CAP_SYS_ADMIN) that isn't available by default... so it's trying to fall back to using the `/bin/fusermount` helper that isn't present.


Try adding `--keep-privs` which is somewhat like using `--privileged` with Docker (which I believe you'd probably need for this). Alternatively use `--add-caps` with the specific capabilities you require.

There is documentation about capabilities for OCI-Mode in the user guide here:

https://docs.sylabs.io/guides/latest/user-guide/security_options.html#oci-mode

**Reply by jordeu** [reply]
Theoretically it shouldn't need `CAP_SYS_ADMIN` because I'm using Kernel 5.15

I've tested adding only `--keep-privs`, adding only `--add-caps CAP_SYS_ADMIN` and adding both, but still I'm not able to mount and I always get same error. 

When using `-B` to mount `/dev/fuse` is exactly the same that is done when not using OCI? or is somehow different?

**Reply by dtrudg** [reply]
It's not the same. `-B /dev/fuse` adds a bind mount definition to the OCI runtime configuration and it is runc/crun that performs the mount. In the native mode, the whole of `/dev` is bound by default by Singularity's own code.

I think I've found a bug in SingularityCE now... I can see that the bind mount under /dev is happening with a `nodev` option - which will obviously cause it to fail. With the native mode, a bind from `/dev` to `/dev` does not set `nodev`.

I'll open an issue and we'll fix this.

**Reply by dtrudg** [reply]
Opened #2326

**Reply by jordeu** [reply]
Thanks!

**Reply by dtrudg** [reply]
@jordeu - if you are able to try PR #2328 that would be appreciated. It works for me...

```
$ singularity run --keep-privs --oci -B .:/root -B /dev/fuse --fakeroot  docker://ubuntu
INFO:    Using cached OCI-SIF image
root@mini:~# ./gocryptfs cipher_oci plain_oci/
Password: 
Decrypting master key
DetectQuirks: Btrfs detected, forcing -noprealloc. See https://github.com/rfjakob/gocryptfs/issues/395 for why.
Filesystem mounted and ready.
```

You will need `--keep-privs` or `--add-caps CAP_SYS_ADMIN`. We're inside a user namespace, but a `mount` call still requires `CAP_SYS_ADMIN` in that namespace (just not outside of it), and it's not provided by OCI-Mode by default.

**Reply by jordeu** [reply]
@dtrudg I've tested current main branch, and now it works as expected. Thank you!

**Reply by dtrudg** [reply]
Great, thanks. 4.0.2 release will happen next week (hopefully) or the week after.

**Reply by marcodelapierre** [reply]
Fantastic Dave, thank you!

**Reply by dtrudg** [reply, marked as answer]
This is now fixed in the CE 4.0.2 release that went out yesterday.

**Reply by marcodelapierre** [reply]
Tested and merged by the team, works successfully with Nextflow+Fusion - thank you!
